### DISCUSSION TITLE
Connecting and disconnecting MIDI pipes

### DISCUSSION THREAD

**Discussion by Vahen1981** [tttapa/Control-Surface#517] [Q&A]
Hi Pieter!
I would like to change states and connect and disconnect two interfaces. I can do it succesfully disconnecting but when I try to connect again it doesn`t work... 
I think I should use one of this methods: 
connectSinkPipe (MIDI_Pipe *sink)
connectSourcePipe (MIDI_Pipe *source)
But, I don't understand what arguments I should pass...

Here is my code:

```
#include <Control_Surface.h>
USBMIDI_Interface midiusb;
HardwareSerialMIDI_Interface midiser = {Serial1, MIDI_BAUD};
MIDI_PipeFactory<3> pipes;

NoteButton panelButtons[]={
  { 2, {MCU::AUTOMATION_TRIM} },
  { 3, {MCU::GROUP} },
};

bool getButton(byte index){
  byte buttonState = panelButtons[index].getButtonState();
  if(buttonState == 0b01){
    return true;}
  else{return false;}}

void setup() {
  midiusb << pipes << midiser;
  Control_Surface >> pipes >> midiusb;
  midiusb >> pipes >> Control_Surface;
  midiusb.begin();
  midiser.begin();
  Control_Surface.begin();
  }

void loop() {
  Control_Surface.loop();
  Estados();
}

#define PAGE1        0
#define PAGE2        1
byte estado = PAGE1;

void ESTADOS(){
  switch(estado){
    
  case PAGE1:
       midiser.disconnectSinkPipes();
       if(getButton(0)){estado = PAGE2; break;}
       break;
   
   case PAGE2:
       midiser.connectSinkPipe(midiusb);
       if(getButton(1)){estado = PAGE1; break;}
       break;
   }
}
```

Thanks!!

**Comment by tttapa** [comment]
Consider the following:
```cpp
USBMIDI_Interface midiusb;
HardwareSerialMIDI_Interface midiser = Serial1;
MIDI_Pipe pipe;
...
midiusb << pipe << midiser;
```
Here, `midiser` is the MIDI source and `midiusb` is the sink, this will be important later.

### Disconnecting two interfaces

If you want to remove this connection, the easiest way is to just disconnect the pipe:
```cpp
pipe.disconnect();
```
Alternatively (e.g. if you didn't keep track of the pipe), you can specify explicitly which interfaces to disconnect from each other:
```cpp
midiser.MIDI_Source::disconnect(midiusb);
```
Note that you have to specify the role of `midiser` in the connection you want to remove: `midiser` was the source in the connection, so we use `midiser.MIDI_Source::disconnect`, and the sink (`midiusb`) is passed as an argument.

You can also do it the other way around:
```cpp
midiusb.MIDI_Sink::disconnect(midiser);
```
Here you indicate that `midiusb` is the sink.  

All three approaches have the same effect.

### Connecting two interfaces

If you want to connect the interfaces again, you do exactly the same thing you did in the setup:
```cpp
midiusb << pipe << midiser;
```
Or the other way around:
```cpp
midiser >> pipe >> midiusb;
```
Note that to connect them, you have to specify which pipe to use. Since you're using a factory, you have to be specific, you don't want a new pipe, you want to reuse the old one:
```cpp
midiusb << pipes[0] << midiser;
```
But it's probably best not to use the factory if you're going to reuse pipes, just keep one pipe separate as in the first snippet in this reply.

### Side notes

`midiusb << pipe << midiser` is just shorthand for
```cpp
midiusb.connectSourcePipe(pipe);
midiser.connectSinkPipe(pipe);
```

**Reply by Vahen1981** [reply]
Thank you very much!
Now works perfect! Your answer was very detailed, and makes me understand very well how pipes works

My control surface looks more and more cool thanks to your library :)

![20210710_175403](https://user-images.githubusercontent.com/77554967/125177256-4c8d0e00-e1a8-11eb-942c-c5f161abebc6.jpg)

**Reply by tttapa** [reply]
Glad to hear!

That controller looks really nice! (Are those motorized faders?)

**Reply by Vahen1981** [reply]
I bought just one, to try, but I still don´t start programing... is possible to see it in the picture in the first channel, and I think in a couple of weeks I will start trying.. so let's see whats happen!

**Reply by tttapa** [reply]
I'm also experimenting with motorized faders right now, I ended up offloading the control loops for the motors to some Arduino Nanos, it was a bit tricky to get the timing right otherwise (especially when reading other analog inputs or driving slow displays). I'll upload my code and findings to this repository when I'm done cleaning up.

**Reply by Vahen1981** [reply]
ohh! that would be great. I'm sure your findings will help a lot... I'm very new in this things, so I think will be very hard for me to make that things work haha, but at least the process is very fun. Are you using one arduino nano per fader?

**Reply by tttapa** [reply]
I've uploaded my code here: https://github.com/tttapa/Control-Surface-Motor-Fader  
and my notes here: https://tttapa.github.io/Pages/Arduino/Control-Theory/Motor-Fader/

Each Nano can drive up to four faders, that's all it has pins for (2×4 for the motors, 4 for the touch sensing, 4 for the potentiometers, and then 2 more for the I²C interface). A fifth fader would also require involving the third hardware timer.

**Reply by Vahen1981** [reply]
This is amazing! thank you, I will chek it out as soon as possible, I'm sure it will be very useful!

**Reply by Vahen1981** [reply]
I started trying your code, but, at least in Motor controller I can´t compile, apparentely because the compiler can´t find AH/Math/Divide.hpp 
Any idea?

**Reply by tttapa** [reply]
Ah, my bad, I added that file to Control Surface recently, it looks like I didn't push it yet, just a moment!

**Reply by tttapa** [reply]
I've pushed the latest version to GitHub, you'll have to upgrade Control Surface to 735eadf9ea4084fc976f3dc5c64090af1554e884.

**Reply by Vahen1981** [reply]
Great! now it works... I have some issue that I'm trying to fix... let´s see if I can... if not I would open a new topic. Thanks a lot!

**Reply by tttapa** [reply]
I've enabled the “Discussions” section over at https://github.com/tttapa/Control-Surface-Motor-Fader/discussions, so if you have any questions or issues feel free to open a thread there.
